pangeo_forge_recipes/transforms.py

```diff
@@ -437,6 +438,8 @@ class WriteCombinedReference(beam.PTransform, ZarrWriterMixin):
     :param output_json_fname: Name to give the output references file. Must end in ``.json``.
     """
 
+    store_name: str
+    target_root: Union[str, FSSpecTarget] = field(default_factory=os.getcwd)
```

Comment:
hmmmmm, I'm not a fan of getcwd as now behavior fundamentally differs based on where we're running this from. This leads to confusing to debug stuff...

I made a different suggestion for fixing this in https://github.com/pangeo-forge/pangeo-forge-recipes/issues/587#issuecomment-1699955200.

Reply:
Fair enough!

What about a sentinel value (based on [this SO](https://stackoverflow.com/a/69239756) and [this PEP](https://peps.python.org/pep-0661/#using-class-objects)), something like:
```python
class InjectableDefault:
   """Serves as a default for values which we expect to be either set or injected."""
   pass

def injectable_default():
    return InjectableDefault()

@dataclass
class StoreToZarr(...):
    ...
    target_root: Union[str, FSSpecTarget, InjectableDefault] = field(default_factory=injectable_default)
```

Reply:
Does that possibly make debugging clearer because it just won't work?

Trying to see if we can add this as a small patch without dropping a whole Python version support...

Reply:
A possible benefit of this `InjectableDefault` approach is that it provides a type hint in the code which can flag a given argument as also being declared in https://github.com/pangeo-forge/pangeo-forge-recipes/pull/566 (once that lands)? So when we read the `transforms` module... we know what we expect to be injected, without cross-referencing to the injection spec added by that PR...?

Reply:
Nah, we should just drop the python version support! If we want to do something with types, we can do that separately from this - but IMO in this case, the right way to tackle this is to drop 3.9 support.

Reply:
> IMO in this case, the right way to tackle this is to drop 3.9 support.

Just realizing that the only thing dropping 3.9 solves is where we set the default, not what the default is.

With 3.10, we can set the default on `ZarrWriterMixin` and still have required fields on child classes. With 3.9, we need to set the default on the child classes.

But in either case, we need to chose _what_ the default is. https://github.com/pangeo-forge/pangeo-forge-recipes/pull/588/commits/5a37cb7a7c5a17412517e176951a97e6a07e3a7d demonstrates the sentinel class idea (again, this could be thought of as agnostic to 3.9 or 3.10, we need a default either way).

I think the empty sentinel works nicely, and indeed prevents the recipe from being run in the first place.

What if we use something like this here, so we can just merge this, and drop 3.9 in a more deliberate PR?